nipype/interfaces/bids_utils.py

```diff
@@ -0,0 +1,169 @@
+                               desc='Queries for outfield outputs')
+    raise_on_empty = traits.Bool(True, usedefault=True,
+                                 desc='Generate exception if list is empty '
+                                 'for a given field')
+    return_type = traits.Enum('file', 'namedtuple', usedefault=True)
+
+
+class BIDSDataGrabber(BaseInterface):
+
+    """ BIDS datagrabber module that wraps around pybids to allow arbitrary
+    querying of BIDS datasets.
+
+    Examples
+    --------
+
+    >>> from nipype.interfaces.bids_utils import BIDSDataGrabber
+    >>> from os.path import basename
+
+    By default, the BIDSDataGrabber fetches anatomical and functional images
+    from a project, and makes BIDS entities (e.g. subject) available for
+    filtering outputs.
+
+    >>> bg = BIDSDataGrabber()
+    >>> bg.inputs.base_dir = 'ds005/'
+    >>> bg.inputs.subject = '01'
+    >>> results = bg.run()
+    >>> basename(results.outputs.anat[0]) # doctest: +ALLOW_UNICODE
+    'sub-01_T1w.nii.gz'
+
+    >>> basename(results.outputs.func[0]) # doctest: +ALLOW_UNICODE
+    'sub-01_task-mixedgamblestask_run-01_bold.nii.gz'
+
+
+    Dynamically created, user-defined output fields can also be defined to
+    return different types of outputs from the same project. All outputs
+    are filtered on common entities, which can be explicitly defined as
+    infields.
+
+    >>> bg = BIDSDataGrabber(infields = ['subject'], outfields = ['dwi'])
+    >>> bg.inputs.base_dir = 'ds005/'
+    >>> bg.inputs.subject = '01'
+    >>> bg.inputs.output_query['dwi'] = dict(modality='dwi')
+    >>> results = bg.run()
+    >>> basename(results.outputs.dwi[0]) # doctest: +ALLOW_UNICODE
+    'sub-01_dwi.nii.gz'
+
+    """
+    input_spec = BIDSDataGrabberInputSpec
+    output_spec = DynamicTraitedSpec
+    _always_run = True
+
+    def __init__(self, infields=None, outfields=None, **kwargs):
+        """
+        Parameters
+        ----------
+        infields : list of str
+            Indicates the input fields to be dynamically created
+
+        outfields: list of str
+            Indicates output fields to be dynamically created.
+            If no matching items, returns Undefined.
+        """
+        super(BIDSDataGrabber, self).__init__(**kwargs)
+        if not have_pybids:
+            raise ImportError("The BIDSEventsGrabber interface requires pybids."
+                              " Please make sure it is installed.")
+
+        # If outfields is None use anat and func as default
+        if outfields is None:
+            outfields = ['func', 'anat']
+            self.inputs.output_query = {
+                "func": {"modality": "func"},
+                "anat": {"modality": "anat"}}
+        else:
+            self.inputs.output_query = {}
+
+        # If infields is None, use all BIDS entities
+        if infields is None:
+            bids_config = join(dirname(gb.__file__), 'config', 'bids.json')
```

Comment:
as a future improvement to a GrabbitIT interface that could serve as a base class, could this be passed along as a parameter?

this would also allow arbitrary and newer layouts to be used.

Reply:
Yes. The issue here is that I cannot get the entities of a project (or BIDS entities in general), without initializing a BIDSLayout with a specific dataset. So this was the only way I could think of to access the entities definition.